bee-common/bee-common/src/packable/mod.rs

```diff
@@ -0,0 +1,207 @@
+            }
+
+            fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
+                let mut bytes = [0u8; core::mem::size_of::<Self>()];
+                unpacker.unpack_bytes(&mut bytes).map_err(UnpackError::Unpacker)?;
+                Ok(Self::from_le_bytes(bytes))
+            }
+
+            fn packed_len(&self) -> usize {
+                core::mem::size_of::<Self>()
+            }
+        }
+    };
+}
+
+impl_packable_for_int!(u8);
+impl_packable_for_int!(u16);
+impl_packable_for_int!(u32);
+impl_packable_for_int!(u64);
+#[cfg(has_u128)]
+impl_packable_for_int!(u128);
+
+/// `usize` integers are packed and unpacked as `u64` integers according to the spec.
+impl Packable for usize {
+    type Error = Infallible;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        (*self as u64).pack(packer)
+    }
+
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
+        Ok(unpacker.unpack_infallible::<u64>().map_err(UnpackError::Unpacker)? as usize)
+    }
+
+    fn packed_len(&self) -> usize {
+        core::mem::size_of::<u64>()
+    }
+}
+
+impl_packable_for_int!(i8);
+impl_packable_for_int!(i16);
+impl_packable_for_int!(i32);
+impl_packable_for_int!(i64);
+#[cfg(has_i128)]
+impl_packable_for_int!(i128);
+
+/// `isize` integers are packed and unpacked as `i64` integers according to the spec.
+impl Packable for isize {
+    type Error = Infallible;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        (*self as i64).pack(packer)
+    }
+
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
+        Ok(unpacker.unpack_infallible::<i64>().map_err(UnpackError::Unpacker)? as isize)
+    }
+
+    fn packed_len(&self) -> usize {
+        core::mem::size_of::<i64>()
+    }
+}
+
+impl Packable for bool {
+    type Error = Infallible;
+
+    /// Booleans are packed as `u8` integers following Rust's data layout.
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        (*self as u8).pack(packer)
+    }
+
+    /// Booleans are unpacked if the byte used to represent them is non-zero.
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
+        Ok(unpacker.unpack_infallible::<u8>().map_err(UnpackError::Unpacker)? != 0)
+    }
+
+    fn packed_len(&self) -> usize {
+        core::mem::size_of::<u8>()
+    }
+}
+
+/// Error type raised when a semantic error occurs while unpacking an option.
+#[derive(Debug)]
+pub enum UnpackOptionError<T: Packable> {
+    /// The tag found while unpacking is not valid.
+    UnknownTag(u8),
+    /// A semantic error for the underlying type was raised.
+    Inner(T::Error),
+}
+
+/// Options are packed and unpacked using `0u8` as the prefix for `None` and `1u8` as the prefix
+/// for `Some`.
+impl<T: Packable> Packable for Option<T> {
```

Comment:
Does deriving not work here?

Reply:
can you derive for types that are defined in `core`/`std`?

Reply:
Fair point, I'm not using my brain today.